### PR TITLE
Feature/9672 special hud

### PR DESCRIPTION
**Pull request by KaiTeuber**
## Description
Added a special HUD to show while scanning for QR Code data.
HUD overlay won't cover the whole screen now so the user can cancel processing

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-9672

## Screenshots
Screen Shot 2021-09-29 at 15 56 45

### CODE REVIEW

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift [hunk 1]
```diff
@@ -12,13 +12,11 @@ class FileScannerCoordinatorViewModel: NSObject, PHPickerViewControllerDelegate,
 	// MARK: - Init
 
 	init(
-		showHUD: @escaping () -> Void,
-		hideHUD: @escaping () -> Void,
+		showHUD: @escaping (@escaping () -> Void) -> Void,
```
Comment (pascal-brause) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift:15]: 🧐

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerHUD.swift [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+class FileScannerHUD {
+
+	// MARK: - Init
+
+	init(
+		execute: @escaping () -> Void
+	) {
+		guard let window = UIApplication.shared.keyWindow else {
+			fatalError("Failed to init - not window found")
```
Comment (pascal-brause) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerHUD.swift:16]: Perhaps use here a less immersive error?

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerHUD.swift [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+import Foundation
+import UIKit
+
+class FileScannerHUD {
+
+	// MARK: - Init
+
+	init(
+		execute: @escaping () -> Void
```
Comment (pascal-brause) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerHUD.swift:13]:
```suggestion
		show: @escaping () -> Void
```
Just an idea...

Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerHUD.swift:13]: execute is the code run while the hud is shown - i thin it better matches

Comment (pascal-brause) [reply src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerHUD.swift:13]: ah then i mixed there something up. Then it's fine for me.

**Review by pascal-brause — APPROVED** [review]

**Review by ChristinaTsigganos — APPROVED** [review]
UA approved

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift [hunk 4]
```diff
@@ -27,46 +25,45 @@ class FileScannerCoordinatorViewModel: NSObject, PHPickerViewControllerDelegate,
...
+		dismiss()
+		showHUD { [weak self] in
+			results.forEach { result in
+				let itemProvider = result.itemProvider
+				guard itemProvider.canLoadObject(ofClass: UIImage.self) else {
+					self?.qrCodesFound([])
 					return
 				}
-				self.qrCodesFound(codes)
+				itemProvider.loadObject(ofClass: UIImage.self) { [weak self]  provider, _ in
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift:36]: This part looks similar to the part in `didFinishPickingMediaWithInfo`. Would be nice to have a new func and avoid code duplication.

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerHUD.swift [hunk 5]
```diff
@@ -0,0 +1,73 @@
+//
+// 🦠 Corona-Warn-App
+//
+
+import Foundation
+import UIKit
+
+class FileScannerHUD {
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerHUD.swift:8]:
```suggestion
class FileScannerLoadingIndicator {
```

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/HUDView.swift [hunk 6]
```diff
@@ -0,0 +1,70 @@
+//
+// 🦠 Corona-Warn-App
+//
+
+import UIKit
+
+class HUDView: UIView {
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/HUDView.swift:7]:
```suggestion
class FileScannerHUDView: UIView {
```

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerHUD.swift [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+	private let window: UIWindow
+	private let hudView: HUDView
+	private let execute: () -> Void
+	private let duration = 0.45
+
+	private func setupView() {
+		hudView.translatesAutoresizingMaskIntoConstraints = false
+		hudView.isUserInteractionEnabled = false
+		window.addSubview(hudView)
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerHUD.swift:60]: I guess there is no other way to add the view to the view hierarchy at the view controller level?
To me it looks hacky to add the view to the window directly (also without a view controller).
Just have questions in mind like: 
What about the view controller lifecycle? The `hudView` is not managed under the lifecycle.
What about the user dismissing the view controller in some untracked way. Will the hud be stuck at the window level?
What about some other logic dismissing the view controller, maybe some system logic or some hidden logic in our code (which maybe will be added at a later time)?

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift [hunk 8]
```diff
@@ -27,46 +25,45 @@ class FileScannerCoordinatorViewModel: NSObject, PHPickerViewControllerDelegate,
...
-			}
-			itemProvider.loadObject(ofClass: UIImage.self) { [weak self]  provider, _ in
-				guard let self = self,
-					  let image = provider as? UIImage,
-					  let codes = self.findQRCodes(in: image),
-					  !codes.isEmpty
-				else {
-					Log.debug("Looks like we have an issue reading the image", log: .fileScanner)
+		dismiss()
+		showHUD { [weak self] in
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift:29]: Why does the processing of the results needs to wait for the hud to show up? There is no user interaction or some other async process in between. At least I don't see it. And all is happening on the main thread anyways.

**Review by ArturFriesen — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerHUD.swift (hunk 7 above)
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerHUD.swift]: Please have a look on the changes I made

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinator.swift [hunk 9]
```diff
@@ -6,31 +6,31 @@ import UIKit
 import PhotosUI
 
 class FileScannerCoordinator {
-
+
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinator.swift:9]: Why are these changes happening?

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift [hunk 10]
```diff
@@ -31,46 +29,45 @@ class FileScannerCoordinatorViewModel: NSObject, PHPickerViewControllerDelegate,
...
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift:33]: Question still valid after refactoring: Why does the processing of the results needs to wait for the hud to show up? There is no user interaction or some other async process in between. At least I don't see it. And all is happening on the main thread anyways.

**Review by ArturFriesen — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift (hunk 10 above)
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift:33]: I made a refactoring to use showHud and hideHud only

**Review by pascal-brause — APPROVED** [review]

**Pull request merged**